### ISSUE TITLE
Bug in DoFHandler and hp::DoFHandler

### ISSUE THREAD

**Issue by luca-heltai** [dealii/dealii#6325]
The following minimal example triggers an exception

~~~
Triangulation<dim,spacedim> tria;
  GridGenerator::hyper_cube (tria);
  FE_Q<dim,spacedim> fe(1);

  DoFHandler<dim,spacedim> dh(tria);
  dh.distribute_dofs(fe);

  // This stores a pointer to the fe in dh.
  FEValues<dim,spacedim> fe_v(dh.get_fe(), QGauss<dim>(1), update_values);

  tria.refine_global (1);
  dh.distribute_dofs(fe);
~~~
The second `distribute_dofs` will fail, saying that someone is still using the fe.

Notice that if we use `fe` instead of `dh.get_fe()`, the test passes.

Should we **clone** the fe also inside `FEValues`?

**Comment by luca-heltai** [comment]
In principle, the code in 

~~~
 if  (fe_collection.size() == 0 || &(fe_collection[0]) != &ff)
   ...
~~~
should prevent this from happening, but this is not working, since fe_collection **clones** the fe, so the two addresses will never be the same, unless i replace the code above with

~~~
Triangulation<dim,spacedim> tria;
  GridGenerator::hyper_cube (tria);
  FE_Q<dim,spacedim> fe(1);

  DoFHandler<dim,spacedim> dh(tria);
  dh.distribute_dofs(fe);

  // This stores a pointer to the fe in dh.
  FEValues<dim,spacedim> fe_v(dh.get_fe(), QGauss<dim>(1), update_values);

  tria.refine_global (1);
  dh.distribute_dofs(dh.get_fe());
~~~
which is not how you would expect this to work...

**Comment by masterleinad** [comment]
Just to be clear: We had this issue also before #6305, don't we?

**Comment by luca-heltai** [comment]
Yes, I think we did. The bug I filed is there also on earlier commits.

**Comment by masterleinad** [comment]
It should be sufficient to check for `FiniteElement::get_name()`. PR in preparation.

**Comment by luca-heltai** [comment]
How about we clone the fe also in FEValues?

**Comment by luca-heltai** [comment]
After all, with the modifications in dofhandler, there is no reason to keep a smartpointer around to FiniteElements anywhere. All of these should really be independent *clones*.

**Comment by luca-heltai** [comment]
(and, of course, checking for the name is also saving a clone if it is not necessary)

**Comment by kronbichler** [comment]
> How about we clone the fe also in FEValues?

Some elements are expensive to set up - essentially all those where we did not set up the lazy initialization of interpolation matrices. I would not be surprised if someone complained about a performance regression. (Not that I disagree with you because I think finite elements should not hold such matrices at all.)

**Comment by luca-heltai** [comment]
But then, why clone them in DoFHandler in the first place? SmartPointers to fe are in FEValues, but also in Mapping (QEulerian and FEField), and I bet somewhere else...

All of these now point to a *different* object, of the same type, and things don't break by chance. 

For example, no code using MappingQEulerian is currently working beyond one refinement.

**Comment by drwells** [comment]
> All of these now point to a different object, of the same type, and things don't break by chance.

I have to confess that I did not consider this case when I pushed for cloning these things.

> For example, no code using MappingQEulerian is currently working beyond one refinement.

This is scary. @luca-heltai do you have a minimum working example?

**Comment by luca-heltai** [comment]
@drwells, yes, the one described above already fails:

~~~
  Triangulation<dim,spacedim> tria;
  GridGenerator::hyper_cube (tria);
  FE_Q<dim,spacedim> fe(1);

  DoFHandler<dim,spacedim> dh(tria);
  dh.distribute_dofs(fe);

  // This stores a pointer to the fe in dh.
  FEValues<dim,spacedim> fe_v(dh.get_fe(), QGauss<dim>(1), update_values);

  tria.refine_global (1); 
  dh.distribute_dofs(fe);  // Throws when trying to destroy the internal fe in dh
~~~

**Comment by luca-heltai** [comment]
There is another test in #6336, and one in #6296, all depending on this bug (which I think will be closed by #6330).

**Comment by bangerth** [comment]
I've been thinking about ways we can break this. One would be that `DoFHandler` uses a `std::shared_ptr` and `FEValues` and everyone else simply signs on to the same `std::shared_ptr`. This way, it will live as long as that is necessary.

But semantically, I think the underlying issue is that in these testcases, there is an object that references a finite element object that is supposed to be part of a `DoFHandler` that now no longer uses it. In fact, one could argue that every `FEValues` object that persists through a call to `DoFHandler::distribute_dofs()` is likely no longer in a valid state since it references a cell that no longer exists. The same may be the case for `MappingFEField`. Could we come up with a signal mechanism that these classes subscribe to in a way where the signal is triggered at the top of `distribute_dofs()`, and these classes then drop their reference to the finite element? In the case of `FEValues`, we could also just trigger an assertion and decide that no such object should live past a call to `distribute_dofs()` -- this would avoid the pain of having to decide what exactly is supposed to happen in this situation: we just declare the testcases as invalid.
